idx/cassandra/cassandra.go

```diff
@@ -236,46 +235,67 @@ func (c *CasIdx) Stop() {
 func (c *CasIdx) AddOrUpdate(data *schema.MetricData, partition int32) idx.Archive {
 	pre := time.Now()
 	existing, inMemory := c.MemoryIdx.Get(data.Id)
-	updateIdx := false
+	archive := c.MemoryIdx.AddOrUpdate(data, partition)
 	stat := statUpdateDuration
+	if !inMemory {
+		stat = statAddDuration
+	}
+	if !updateCassIdx {
+		stat.Value(time.Since(pre))
+		return archive
+	}
 
-	if inMemory {
-		if existing.Partition == partition {
-			var oldest time.Time
-			if updateInterval > 0 {
-				oldest = time.Now().Add(-1 * updateInterval).Add(-1 * time.Duration(rand.Int63n(updateInterval.Nanoseconds()*int64(updateFuzzyness*100)/100)))
-			} else {
-				oldest = time.Now()
-			}
-			updateIdx = (existing.LastUpdate < oldest.Unix())
-		} else {
-			if updateCassIdx {
-				// the partition of the metric has changed. So we need to delete
-				// the current metricDef from cassandra.  We do this in a separate
-				// goroutine as we dont want to block waiting for the delete to succeed.
-				go func() {
-					if err := c.deleteDef(&existing); err != nil {
-						log.Error(3, err.Error())
-					}
-				}()
+	now := uint32(time.Now().Unix())
+
+	// Cassandra uses partition id asthe partitionin key, so an "update" that changes the partition for
+	// an existing metricDef will just create a new row in the table and wont remove the old row.
+	// So we need to explicitly delete the old entry.
+	if inMemory && existing.Partition != partition {
+		go func() {
+			if err := c.deleteDef(&existing); err != nil {
+				log.Error(3, err.Error())
 			}
-			updateIdx = true
-		}
-	} else {
-		updateIdx = true
-		stat = statAddDuration
+		}()
 	}
 
-	if updateIdx {
-		archive := c.MemoryIdx.AddOrUpdate(data, partition)
-		if updateCassIdx {
-			log.Debug("cassandra-idx updating def in index.")
-			c.writeQueue <- writeReq{recvTime: time.Now(), def: &archive.MetricDefinition}
-		}
+	// check if we need to save to cassandra.
+	if archive.LastSave >= (now - updateInterval32) {
 		stat.Value(time.Since(pre))
 		return archive
 	}
-	return existing
+
+	// This is just a safety precaution to prevent corrupt index entries.
+	// This ensures that the index entry always contains the correct metricDefinition data.
+	if inMemory {
+		archive.MetricDefinition = *schema.MetricDefinitionFromMetricData(data)
+		archive.MetricDefinition.Partition = partition
+	}
+
+	// if the entry has not been saved for 1.5x updateInterval
+	// then perform a blocking save. (bit shifting to the right 1 bit, divides by 2)
+	if archive.LastSave < (now - updateInterval32 - (updateInterval32 >> 1)) {
+		log.Debug("cassandra-idx updating def in index.")
+		c.writeQueue <- writeReq{recvTime: time.Now(), def: &archive.MetricDefinition}
+		archive.LastSave = now
+		c.MemoryIdx.Update(archive)
```

Comment:
i added a counter idx.cassandra.save.skipped to keep track of how many saves are being skipped due to the writeQ being full.  Spikes in this counter would be normal, but continued growth over an extended time would indicate a performance problem.